[macOS] Implement _isInWindowActive SPI

jernoble opened the pull request:
#### 402f8772f2520846e873094226fe4bdb43be14d3
<pre>
[macOS] Implement _isInWindowActive SPI
<a href="https://bugs.webkit.org/show_bug.cgi?id=269499">https://bugs.webkit.org/show_bug.cgi?id=269499</a>
<a href="https://rdar.apple.com/122827633">rdar://122827633</a>

Reviewed by Andy Estes.

Implement _isInWindowActive by pushing activation state up from the WebContent
process to be cached locally in the UIProcess for ease of querying.

- Rename isInWindow() to isInWindowFullscreen() in a number of places,
  for disambiguation reasons.

- Add Debug menu items to MiniBrowser to toggle both in-window fullscreen
  and picture-in-picture modes for testing purposes.

- Add API tests for both availability and active states of in-window fullscreen.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModelClient::isInWindowFullscreenActiveChanged):
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::updateForEventName):
(WebCore::PlaybackSessionModelMediaElement::toggleInWindowFullscreen):
(WebCore::PlaybackSessionModelMediaElement::observedEventNames):
(WebCore::PlaybackSessionModelMediaElement::isInWindowFullscreenActive const):
(WebCore::PlaybackSessionModelMediaElement::toggleInWindow): Deleted.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.h:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.mm:
(WebCore::PlaybackSessionInterfaceMac::isInWindowFullscreenActive const):
(WebCore::PlaybackSessionInterfaceMac::toggleInWindowFullscreen):
(WebCore::PlaybackSessionInterfaceMac::setPlayBackControlsManager):
* Source/WebCore/platform/mac/WebPlaybackControlsManager.mm:
(-[WebPlaybackControlsManager toggleInWindow]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _isInWindowActive]):
(-[WKWebView _toggleInWindow]):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::toggleInWindowFullscreen):
(WebKit::PlaybackSessionModelContext::isInWindowFullscreenActiveChanged):
(WebKit::PlaybackSessionManagerProxy::isInWindowFullscreenActiveChanged):
(WebKit::PlaybackSessionModelContext::toggleInWindow): Deleted.
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController setInWindowFullscreenActive:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::isInWindowFullscreenActive const):
(WebKit::WebViewImpl::toggleInWindowFullscreen):
(WebKit::WebViewImpl::toggleInWindow): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionInterfaceContext::isInWindowFullscreenActiveChanged):
(WebKit::PlaybackSessionManager::isInWindowFullscreenActiveChanged):
(WebKit::PlaybackSessionManager::toggleInWindow):
* Tools/MiniBrowser/mac/BrowserWindowController.h:
* Tools/MiniBrowser/mac/BrowserWindowController.m:
(-[BrowserWindowController togglePictureInPicture:]):
(-[BrowserWindowController toggleInWindowFullscreen:]):
* Tools/MiniBrowser/mac/MainMenu.xib:
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController togglePictureInPicture:]):
(-[WK2BrowserWindowController toggleInWindowFullscreen:]):
(-[WK2BrowserWindowController validateMenuItem:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/InWindowFullscreen.mm: Added.
(TestWebKitAPI::createInWindowFullscreenWebView):
(TestWebKitAPI::testEventually):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/274832@main">https://commits.webkit.org/274832@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b95e389173d5e23730775d5fff522bea47124ed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40107 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19119 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42484 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42652 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42414 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22033 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16448 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33367 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40681 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16101 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34641 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13924 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13962 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43930 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36387 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35929 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39685 "Found 1 new test failure: fast/repaint/animation-after-layer-scroll.html (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14922 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12245 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16541 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9010 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16590 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16185 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
```diff
@@ -109,6 +109,7 @@ class PlaybackSessionModel : public CanMakeWeakPtr<PlaybackSessionModel> {
     virtual double volume() const = 0;
     virtual bool isPictureInPictureSupported() const = 0;
     virtual bool isPictureInPictureActive() const = 0;
+    virtual bool isInWindowFullscreenActive() const { return false; }
```
Ditto about `In`

eric-carlson commented on Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
```diff
@@ -136,6 +137,7 @@ class PlaybackSessionModelClient : public CanMakeWeakPtr<PlaybackSessionModelCli
     virtual void volumeChanged(double) { }
     virtual void isPictureInPictureSupportedChanged(bool) { }
     virtual void pictureInPictureActiveChanged(bool) { }
+    virtual void isInWindowFullscreenActiveChanged(bool) { }
```
Ditto here and all the other similar names

eric-carlson commented on Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
```diff
@@ -72,7 +72,7 @@ class PlaybackSessionModel : public CanMakeWeakPtr<PlaybackSessionModel> {
     virtual void selectLegibleMediaOption(uint64_t index) = 0;
     virtual void togglePictureInPicture() = 0;
-    virtual void toggleInWindow() = 0;
+    virtual void toggleInWindowFullscreen() = 0;
```
The `In` in the name doesn’t seem necessary. How about `toggleWindowFullscreen`?

jernoble replied:
The "in" is "in"portant. :) 

This is about entering an in-window fullscreen, so dropping the "in" would change the meaning.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -452,6 +452,14 @@ - (void)setPictureInPictureActive:(BOOL)active
+- (void)setInWindowFullscreenActive:(BOOL)active
+{
+    if (_inWindowFullscreenActive == active)
+        return;
```
Nit: not sure this check buys anything

jernoble replied:
Yeah, I actually just got rid of this change entirely.

aestes commented on Tools/TestWebKitAPI/Tests/mac/InWindowFullscreen.mm:
```diff
@@ -0,0 +1,141 @@
+{
+    RetainPtr configuration = [WKWebViewConfiguration _test_configurationWithTestPlugInClassName:@"WebProcessPlugInWithInternals" configureJSCForTesting:YES];
+    [configuration preferences].elementFullscreenEnabled = YES;
+    [configuration setMediaTypesRequiringUserActionForPlayback: WKAudiovisualMediaTypeAudio];
```
Nit: extra space after :

jernoble committed:
[macOS] Implement _isInWindowActive SPI
https://bugs.webkit.org/show_bug.cgi?id=269499
rdar://122827633

Reviewed by Andy Estes.

Implement _isInWindowActive by pushing activation state up from the WebContent
process to be cached locally in the UIProcess for ease of querying.

- Rename isInWindow() to isInWindowFullscreen() in a number of places,
  for disambiguation reasons.

- Add Debug menu items to MiniBrowser to toggle both in-window fullscreen
  and picture-in-picture modes for testing purposes.

- Add API tests for both availability and active states of in-window fullscreen.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModelClient::isInWindowFullscreenActiveChanged):
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::updateForEventName):
(WebCore::PlaybackSessionModelMediaElement::toggleInWindowFullscreen):
(WebCore::PlaybackSessionModelMediaElement::observedEventNames):
(WebCore::PlaybackSessionModelMediaElement::isInWindowFullscreenActive const):
(WebCore::PlaybackSessionModelMediaElement::toggleInWindow): Deleted.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.h:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.mm:
(WebCore::PlaybackSessionInterfaceMac::isInWindowFullscreenActive const):
(WebCore::PlaybackSessionInterfaceMac::toggleInWindowFullscreen):
(WebCore::PlaybackSessionInterfaceMac::setPlayBackControlsManager):
* Source/WebCore/platform/mac/WebPlaybackControlsManager.mm:
(-[WebPlaybackControlsManager toggleInWindow]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _isInWindowActive]):
(-[WKWebView _toggleInWindow]):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::toggleInWindowFullscreen):
(WebKit::PlaybackSessionModelContext::isInWindowFullscreenActiveChanged):
(WebKit::PlaybackSessionManagerProxy::isInWindowFullscreenActiveChanged):
(WebKit::PlaybackSessionModelContext::toggleInWindow): Deleted.
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController setInWindowFullscreenActive:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::isInWindowFullscreenActive const):
(WebKit::WebViewImpl::toggleInWindowFullscreen):
(WebKit::WebViewImpl::toggleInWindow): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionInterfaceContext::isInWindowFullscreenActiveChanged):
(WebKit::PlaybackSessionManager::isInWindowFullscreenActiveChanged):
(WebKit::PlaybackSessionManager::toggleInWindow):
* Tools/MiniBrowser/mac/BrowserWindowController.h:
* Tools/MiniBrowser/mac/BrowserWindowController.m:
(-[BrowserWindowController togglePictureInPicture:]):
(-[BrowserWindowController toggleInWindowFullscreen:]):
* Tools/MiniBrowser/mac/MainMenu.xib:
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController togglePictureInPicture:]):
(-[WK2BrowserWindowController toggleInWindowFullscreen:]):
(-[WK2BrowserWindowController validateMenuItem:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/InWindowFullscreen.mm: Added.
(TestWebKitAPI::createInWindowFullscreenWebView):
(TestWebKitAPI::testEventually):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/274832@main
